### DISCUSSION TITLE
How is history back handled internally?

### DISCUSSION THREAD

**Discussion by josephernest** [unpoly/unpoly#405] [General]
I'm discovering unpoly, which looks great! Out of curiosity, how is "history back" handled? (if you can give the big picture story @triskweline)

When rendering a page `index.html` for the first time, how does it know how it should be rendered again after a navigation and a history back / `onpopstate`?
How does Unpoly know which "fragment" will be used when going back to `index.html`: should everything be rendered? Or only `<div class="content">`, or something else? What tells which part should be restored, and which part shouldn't (to avoid reload everything / page flickering) which doing "history back"?

Also, does unpoly save/back up the DOM state in localStorage (or similar) for every page of the navigation, so that it can be restored later? (I think HTMX does this, but I need to check)

Thanks!

**Comment by triskweline** [comment]
By default Unpoly will render the `<body>` element from the previous URL. You can configure different targets in [`up.history.config.restoreTargets`](https://unpoly.com/up.history.config#config.restoreTargets).

> Also, does unpoly save/back up the DOM state in localStorage (or similar) for every page of the navigation, so that it can be restored later?

While the response for the previous page may be cached, all DOM elements are re-created from the reponse's HTML body.

**Reply by josephernest** [reply]
Thanks a lot! So by default, history back button will reload the whole `<body>` which might indtroduce flickering, and if we want to prevent this (example: we don't want to repaint header / navigation menu / footer), we configure `https://unpoly.com/up.history.config#config.restoreTargets` to our main `#container` which is the only div that should be repainted? Is this correct?

**Comment by triskweline** [comment]
> we configure https://unpoly.com/up.history.config#config.restoreTargets to our main #container which is the only div that should be repainted? Is this correct?

Yes.

Since the repaint is done in single frame you won't see flickering either way. This is really an optimization to update fewer elements or to preserve elements with local state (e.g. a chat box in the application layout).
